diff --git a/prebuilts/api/31.0/public/hal_neverallows.te b/prebuilts/api/31.0/public/hal_neverallows.te
index 105689b8a..d7dc6baaf 100644
--- a/prebuilts/api/31.0/public/hal_neverallows.te
+++ b/prebuilts/api/31.0/public/hal_neverallows.te
@@ -2,6 +2,7 @@
 # network capabilities
 neverallow {
   halserverdomain
+  -hal_keymaster_server
   -hal_bluetooth_server
   -hal_can_controller_server
   -hal_wifi_server
@@ -19,6 +20,7 @@ neverallow {
 # will result in CTS failure.
 neverallow {
   halserverdomain
+  -hal_keymaster_server
   -hal_automotive_socket_exemption
   -hal_can_controller_server
   -hal_tetheroffload_server
diff --git a/public/hal_neverallows.te b/public/hal_neverallows.te
index 105689b8a..d7dc6baaf 100644
--- a/public/hal_neverallows.te
+++ b/public/hal_neverallows.te
@@ -2,6 +2,7 @@
 # network capabilities
 neverallow {
   halserverdomain
+  -hal_keymaster_server
   -hal_bluetooth_server
   -hal_can_controller_server
   -hal_wifi_server
@@ -19,6 +20,7 @@ neverallow {
 # will result in CTS failure.
 neverallow {
   halserverdomain
+  -hal_keymaster_server
   -hal_automotive_socket_exemption
   -hal_can_controller_server
   -hal_tetheroffload_server
